lib/web/sessions.go

```diff
@@ -497,148 +552,337 @@ func (s *sessionCache) ValidateTrustedCluster(validateRequest *auth.ValidateTrus
-			defer c.Close()
-			return out, nil
+func (c *sessionContext) Close() error {
+	closers := c.transferClosers()
+	for _, closer := range closers {
+		c.log.Debugf("Closing %v.", closer)
+		closer.Close()
+	}
+	return nil
+}
+
+type sessionContext struct {
+	log logrus.FieldLogger
+
+	mu      sync.Mutex
+	closers []io.Closer
+}
+
+// addClosers adds the specified closers to this context
+func (c *sessionContext) addClosers(closers ...io.Closer) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.closers = append(c.closers, closers...)
+}
+
+// removeCloser removes the specified closer from this context
+func (c *sessionContext) removeCloser(closer io.Closer) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for i, cls := range c.closers {
+		if cls == closer {
+			c.closers = append(c.closers[:i], c.closers[i+1:]...)
+			return
 		}
-		return nil, trace.Wrap(err)
 	}
-	return out, nil
 }
 
-func (s *sessionCache) SetSession(w http.ResponseWriter, user, sid string) error {
-	d, err := EncodeCookie(user, sid)
+func (c *sessionContext) transferClosers() []io.Closer {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	closers := c.closers
+	c.closers = nil
+	return closers
+}
+
+// waitForWebSession will block until the requested web session shows up in the
+// cache or a timeout occurs.
+func (h *Handler) waitForWebSession(ctx context.Context, req types.GetWebSessionRequest) error {
+	_, err := h.cfg.AccessPoint.GetWebSession(ctx, req)
+	if err == nil {
+		return nil
+	}
+	logger := h.log.WithField("req", req)
+	if !trace.IsNotFound(err) {
+		logger.WithError(err).Debug("Failed to query web session.")
+	}
+	// Establish a watch.
+	watcher, err := h.cfg.AccessPoint.NewWatcher(ctx, services.Watch{
+		Name: teleport.ComponentWebProxy,
+		Kinds: []services.WatchKind{
+			{
+				Kind:    services.KindWebSession,
+				SubKind: services.KindWebSession,
+			},
+		},
+		MetricComponent: teleport.ComponentWebProxy,
+	})
 	if err != nil {
-		return err
+		return trace.Wrap(err)
 	}
-	c := &http.Cookie{
-		Name:     "session",
-		Value:    d,
-		Path:     "/",
-		HttpOnly: true,
-		Secure:   true,
+	defer watcher.Close()
+	matchEvent := func(event services.Event) (services.Resource, error) {
+		if event.Type == backend.OpPut &&
+			event.Resource.GetKind() == services.KindWebSession &&
+			event.Resource.GetName() == req.SessionID {
+			return event.Resource, nil
+		}
+		return nil, trace.CompareFailed("not match")
 	}
-	http.SetCookie(w, c)
-	return nil
+	_, err = waitForResource(ctx, watcher, eventMatcherFunc(matchEvent), h.clock)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to wait for web session.")
+	}
+	return trace.Wrap(err)
 }
 
-func (s *sessionCache) ClearSession(w http.ResponseWriter) {
-	http.SetCookie(w, &http.Cookie{
-		Name:     "session",
-		Value:    "",
-		Path:     "/",
-		HttpOnly: true,
-		Secure:   true,
-	})
+func waitForResource(ctx context.Context, watcher services.Watcher, m eventMatcher, clock clockwork.Clock) (services.Resource, error) {
```

Comment:
This feels like it doesn't really belong in the web package TBH, maybe move this somewhere to services closer to the `Watcher`, along with `eventMatcher`?